File: 19_remove_nth_node_from_end_of_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    int listSize(ListNode *head)
    {
        if (!head->next)
        {
            return 1;
        }

        return 1 + listSize(head->next);
    }

    ListNode *removeNthFromEnd(ListNode *head, int n)
    {
        int size = listSize(head);

        if (size == 1 && n == 1)
        {
            return nullptr;
        }

        if (size == 2 && n == 1)
        {
            head->next = nullptr;
            return head;
        }

        if (size == n)
        {
            return head->next;
        }

        ListNode *dummy = head;

        for (int i = 0; i < size - n - 1; i++)
        {
            dummy = dummy->next;
        }

        dummy->next = dummy->next->next;
        return head;
    }
};

File: medium/19_remove_nth_node_from_end_of_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head.next:
            return None

        if not head.next.next:
            if n == 1:
                head.next = None
                return head
            elif n == 2:
                return head.next
        
        list1 = head
        list2 = head

        for i in range(n):
            list2 = list2.next
        
        if list2 == None:
            return head.next

        while list2.next:
            list1 = list1.next
            list2 = list2.next
        
        print(list1.val)
        print(list2.val)

        list1.next = list1.next.next

        return head